### PR TITLE
added learnable fourier features

### PR DESCRIPTION
**Pull request by popcornell**
I added a simple implementation of https://arxiv.org/pdf/2106.02795.pdf for 1-D dimensional sequences. 
It is basically a learnt variation of sinusoidal positional encodings for transformers

### CODE REVIEW

**Comment by popcornell** [comment]
Here is a plot of the positional encoding features: 

Figure_1

**Comment by sw005320** [comment]
Can you apply `black`?
https://github.com/espnet/espnet/runs/4652302934?check_suite_focus=true#step:8:9

**Comment by sw005320** [comment]
@pengchengguo, could you review this PR?

@popcornell, can you merge the latest master?
We fix the CI issue.

**Comment by popcornell** [comment]
I should have merged latest changes from master.

**Review by pengchengguo — COMMENTED** [review]
Thanks, @popcornell,

The implementation is good to me. I only have some minor suggestions.

**Comment by pengchengguo** [comment]
Besides, could you please add ci tests for this new module, you can refer to `espnet/test/test_positional_encoding.py`.
Do you have the results of learnable fourier features?

File: espnet/nets/pytorch_backend/transformer/embedding.py [hunk 1]
```diff
@@ -127,6 +127,103 @@ def forward(self, x):
...
+        ), "d_model should be divisible by two in order to use this layer."
+        self.w_r = torch.nn.Parameter(torch.empty(1, d_model // 2))
+        self._reset()  # init the weights
+
+        self.hidden_dim = hidden_dim
+        if self.hidden_dim is not None:
+            self.mlp = torch.nn.Sequential(
+                torch.nn.Linear(d_model, hidden_dim, bias=False),
+                torch.nn.GELU(),
+                torch.nn.Linear(hidden_dim, d_model, bias=False),
```
[the remark this answers is not in the document]
Comment (popcornell) [reply espnet/nets/pytorch_backend/transformer/embedding.py:186]: you are right, I only looked at Eq.6 where strangely bias is missing

**Comment by popcornell** [comment]
> Do you have the results of learnable fourier features?
Nope but would be interesting to try them out

**Comment by popcornell** [comment]
@pengchengguo do you know why some errors are not catched when I run `./ci/test_python.sh`  locally ? What is the optimal testing procedure for ESPNet ?

**Comment by ftshijt** [comment]
Hi @popcornell ,  thanks for your contribution. You can set up the CI following instructions here https://github.com/espnet/espnet/blob/master/CONTRIBUTING.md

**Comment by popcornell** [comment]
Okay I think I have to force flake8 with --verbose flag. Running ./ci/test_python.sh does not do it for me, everything seems okay with that but flake8 clearly fails on GitHub checks.

**Comment by ftshijt** [comment]
We have fixed the CI with the latest master branch, can you merge that to fix the current CI issue? Many thanks

**Comment by popcornell** [comment]
Done !

**Review by pengchengguo — APPROVED** [review]

**Comment by ftshijt** [comment]
Could you pull the latest master in espnet so as to fix the broken CI?

**Comment by popcornell** [comment]
closed the PR by error while trying to change the base back and fort (sometimes GitHub diffs screw up and you have to do this trick). Can you reopen it ?

**Comment by sw005320** [comment]
I'm not sure but this PR is for `espnet:gh-pages` instead of `espnet:master` and I cannot reopen it probably due to that.
Sorry to have various troubles.

**Comment by popcornell** [comment]
going to reopen another one then.

**Pull request closed without merging**